chainerx_cc/chainerx/routines/math.cc

```diff
@@ -546,16 +555,23 @@ Array Log(const Array& x) {
 }
 
 Array LogSumExp(const Array& x, const OptionalAxes& axis, bool keepdims) {
+    // TODO(imanishi): Support dtype promotion for arithmetics functions
+    const Array& x_cast = GetKind(x.dtype()) == DtypeKind::kFloat ? x : x.AsType(Dtype::kFloat32);
```

Comment:
How about using `GetMathResultDtype`?

Reply:
`const Array& x_cast = x.AsType(GetMathResultType(x.dtype()), false)` makes a redundant copy.

Reply:
How about this:
```c++
Dtype out_dtype = GetMathResultType(x.dtype());
const Array& x_cast = x.dtype() == out_dtype ? x : x.AsType(out_dtype);
```

Reply:
Fixed.